tasks/pdp/task_prove.go

```diff
@@ -0,0 +1,758 @@
+			}
+			return partialTreeList[i].ElemOffset < partialTreeList[j].ElemOffset
+		})
+		log.Infow("partialTree", "partialTree", partialTreeList)
+	}
+
+	challLevel := proof.NodeLevel(challSubRoot.SubrootSize/LeafSize, arity)
+	challOffset := (challSubRoot.SubrootOffset / LeafSize) >> uint(challLevel-1)
+
+	log.Infow("challSubRoot", "challSubRoot", challSubrootIdx, "challLevel", challLevel, "challOffset", challOffset)
+
+	challSubtreeLeaf := partialTree[elemIndex{Level: challLevel, ElemOffset: challOffset}]
+	if challSubtreeLeaf.Hash != subrootProof.Root {
+		return contract.PDPVerifierProof{}, xerrors.Errorf("subtree root doesn't match partial tree leaf, %x != %x", challSubtreeLeaf.Hash, subrootProof.Root)
+	}
+
+	var out contract.PDPVerifierProof
+	copy(out.Leaf[:], subrootProof.Leaf[:])
+	out.Proof = append(out.Proof, subrootProof.Proof...)
+
+	currentLevel := challLevel
+	currentOffset := challOffset
+
+	for currentLevel < rootLevel {
+		siblingOffset := currentOffset ^ 1
+
+		// Retrieve sibling hash from partialTree or use zero hash
+		siblingIndex := elemIndex{Level: currentLevel, ElemOffset: siblingOffset}
+		siblingElem, ok := partialTree[siblingIndex]
+		if !ok {
+			return contract.PDPVerifierProof{}, xerrors.Errorf("missing sibling at level %d, offset %d", currentLevel, siblingOffset)
+		}
+
+		log.Infow("siblingElem", "siblingElem", siblingElem, "siblingIndex", siblingIndex, "currentLevel", currentLevel, "currentOffset", currentOffset, "siblingOffset", siblingOffset)
+
+		// Append the sibling's hash to the proof
+		out.Proof = append(out.Proof, siblingElem.Hash)
+
+		// Move up to the parent node
+		currentOffset = currentOffset / arity
+		currentLevel++
+	}
+
+	log.Infow("proof complete", "proof", out)
+
+	rootCid, err := cid.Parse(subroots[0].Root)
+	if err != nil {
+		return contract.PDPVerifierProof{}, xerrors.Errorf("failed to parse root CID: %w", err)
+	}
+	commRoot, err := commcid.CIDToPieceCommitmentV1(rootCid)
+	if err != nil {
+		return contract.PDPVerifierProof{}, xerrors.Errorf("failed to convert CID to piece commitment: %w", err)
+	}
+	var cr [LeafSize]byte
+	copy(cr[:], commRoot)
+
+	if !Verify(out, cr, uint64(challengedLeaf)) {
+		return contract.PDPVerifierProof{}, xerrors.Errorf("proof verification failed")
+	}
+
+	// Return the completed proof
+	return out, nil
+}
+
+func (p *ProveTask) getSenderAddress(ctx context.Context) (common.Address, error) {
+	// todo do based on proofset
+
+	var addressStr string
+	err := p.db.QueryRow(ctx, `SELECT address FROM eth_keys WHERE role = 'pdp' LIMIT 1`).Scan(&addressStr)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return common.Address{}, errors.New("no sender address with role 'pdp' found")
+		}
+		return common.Address{}, err
+	}
+	address := common.HexToAddress(addressStr)
+	return address, nil
+}
+
+func (p *ProveTask) cleanupDeletedRoots(ctx context.Context, proofSetID int64, pdpVerifier *contract.PDPVerifier) error {
```

Comment:
There is also the pdp_piecerefs table, and I'm not entirely sure what's the best way to approach this because:
* On upload the piece is only referred in pdp_piecerefs (+the piecepark tables)
* Only after upload is complete and data is verified the PDP service can add the piece to a proofset
* Technically one piece can be in multiple proofsets / roots (as a subroot)
* So probably:
  * Add a flag to pdp_piecerefs which is optionally settable through the delete api which tells curio to drop the piece reference when all pdp_proofset_roots are gone
  * Possibly add an endpoint which lists all pdp_proofset_roots for a given piece (so the service can know why a given piece may be "locked")